### DISCUSSION TITLE
How to get the input control matrix from SOFA

### DISCUSSION THREAD

**Discussion by xlsufo** [sofa-framework/sofa#5007] [SofaPython3] [answered]
Hello everyone！
I have read the paper **FEM Based Workspace Estimation for Soft Robots: a  Forward-Backward Interval Analysis Approach**, which studies a soft robot driven by cables, as shown below.
![Quicker_20240919_092359](https://github.com/user-attachments/assets/f0373acb-fc55-44d8-aff5-81832783f1be)

The Euler-Lagrange equation are constructed

![40919092406](https://github.com/user-attachments/assets/6f78ac75-c960-4ef1-b198-5d5fae7f8314)

I want to reproduce the content of the paper. But I don't know how to get the H matrice from SOFA. 
The paper mentions the reference "The finite element method in engineering", but I don't have the license for this book.
Can anyone help?

**Comment by alxbilger** [comment]
The matrix you want to get is not available through Python in SOFA. There is a work in progress so it becomes available. Stay tune, it may come in the next days/weeks (but you will have to compile SOFA or wait the next release in december).

The very short term solution would be to get it in C++

**Reply by xlsufo** [reply]
Thank you, may I ask if there is some reference for implementation in C++.

**Comment by alxbilger** [comment, marked as answer]
Actually, I realized that the matrix is already available in the Data `constraint` of a `MechanicalObject`. The problem is that it will be a string, and you will have to parse it (making the simulation ultra-slow). Here is an example of what you would get:

![image](https://github.com/user-attachments/assets/b4a7693a-ba6d-4919-825e-51fb7d865044)

**Reply by xlsufo** [reply]
Hello!
I extracted this data from the DiamondRobot in softrobot tutorials.
![Quicker_20240920_173107](https://github.com/user-attachments/assets/d24d6c58-5539-46cd-bbc4-cf14030523a2)

 I reduced the number of cables to one and got the following result:
```
 Constraint ID : 0  
dof ID : 444  value : 0.0004578 -0.0500435 -0.0167298
dof ID : 483  value : 0.00613595 -0.670739 -0.224232
dof ID : 1237  value : 0.00355964 -0.389115 -0.130083
dof ID : 1445  value : -0.00147765 0.161526 0.0539992
```
Is this the H matrix I need? I noticed that my “dof ID” number is 4, while your example is 3. What is the reason for this difference? Maybe because the DiamondRobot  is composed of tetrahedral units?

**Reply by alxbilger** [reply]
you're right. The cables are defined in a different space than the main DoFs. The `BarycentricMapping` maps values from this space to the main DoFs using the topology. In this case, a point in a tetrahedron. This point is distributed to the main DoFs using the barycentric coordinates.

**Reply by xlsufo** [reply]
I'm very sorry that I haven't studied these theories. Based on your explanation, I think that when the tension of the cable is u, the values ​​of above points are a1, a2, a3, and a4. If I want to get the actual force distributed on the four points, can I follow the following formula?
![QianJianTec1726841514566](https://github.com/user-attachments/assets/fe2e1daa-5406-4030-bfa6-e4b07b850c2d)

**Comment by alxbilger** [comment]
A Pull Request has been submitted https://github.com/sofa-framework/SofaPython3/pull/459. If you build SOFA yourself, you can already checkout the branch (but changes may still occur), or you wait that it is merged into `master`. If you use the binaries, you will have to wait for the next release.
